### PR TITLE
Feature id fixes

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes

Adds option to all allow override of the name of the seed column. By default, the seed column is called 'id' - but this can conflict if the name `id` is used as the name of a column with a different function in the generated data. This is accomplished by adding an additional parameter to the creation of the dataspec - `seedColumnName`

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Change to tutorials, tests or examples
- [ ] Non code change (readme, images or other non-code assets)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [x] Submission does not increase alerts or messages from LGTM

## Further comments

Fixes collisions between system 'id' column and user explicitly adding id column by offering parameter to DataGenerator constructor allowing specification of an alternate id column

### CODE REVIEW

**Comment by renardeinside** [comment]
@ronanstokes-db 
seems like you've copied the ISSUE_TEMPLATE without changing it from `dbx` 😄 :

> What types of changes does your code introduce to dbx?

File: dbldatagen/column_generation_spec.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
     TimestampType, DataType, DateType
 
 from .column_spec_options import ColumnSpecOptions
-from .datagen_constants import RANDOM_SEED_FIXED, RANDOM_SEED_HASH_FIELD_NAME, RANDOM_SEED_RANDOM
+from .datagen_constants import RANDOM_SEED_FIXED, RANDOM_SEED_HASH_FIELD_NAME, RANDOM_SEED_RANDOM, DEFAULT_SEED_COLUMN
```
Comment (renardeinside) [inline dbldatagen/column_generation_spec.py:19]: I would recommend applying [`black`](https://github.com/psf/black) formatting globally to the project, this is hard to read

File: dbldatagen/column_generation_spec.py [hunk 2]
```diff
@@ -89,7 +87,9 @@ class ColumnGenerationSpec(object):
 
     def __init__(self, name, colType=None, minValue=0, maxValue=None, step=1, prefix='', random=False,
                  distribution=None, baseColumn=None, randomSeed=None, randomSeedMethod=None,
-                 implicit=False, omit=False, nullable=True, debug=False, verbose=False, **kwargs):
+                 implicit=False, omit=False, nullable=True, debug=False, verbose=False,
```
Comment (renardeinside) [inline dbldatagen/column_generation_spec.py:90]: This is hard to maintain/develop. Probably these properties relate to various behavioural objects. 
You can separate them using pydantic or dataclass, for instance:

```python

from pydantic import BaseModel

class PropertySetNumberOne(BaseModel):
    col_type: Optional[str] = None
    min_value: Optional[int] = 1
 
 class PropertySetNumberOne:
    implicit: bool = False
    omit: bool = False
    
 class MyNiceClass: 
    ....
    # add here implementation for .with methods
 
# example class invocation

obj_instance = (
    MyNiceClass
        .withPropertySet(PropertySet1())
        .withAnotherPropertySet(PropertySet2())
)
```

Also the variable naming is inconsistent - input parameters are `camelCase`, but default is snake_case (e.g. `DEFAULT_SEED_COLUMN`.

File: dbldatagen/data_generator.py [hunk 3]
```diff
@@ -257,7 +274,8 @@ def explain(self, suppressOutput=False):
 
         output = ["", "Data generation plan", "====================",
```
Comment (renardeinside) [inline dbldatagen/data_generator.py:275]: this `"===================="` is not aligned to the screen width. Use something like `rich` or `colorama` for proper printing.

File: dbldatagen/column_generation_spec.py (hunk 1 above)
Comment (ronanstokes-db) [reply dbldatagen/column_generation_spec.py:19]: interesting suggestion but more on medium term task list

File: dbldatagen/column_generation_spec.py (hunk 2 above)
Comment (ronanstokes-db) [reply dbldatagen/column_generation_spec.py:90]: Agreed but not part of code changes for this release

Actually code does use options object but as this was used in some high value deals even in prerelease version, did not want to change API signatures

File: dbldatagen/data_generator.py (hunk 3 above)
Comment (ronanstokes-db) [reply dbldatagen/data_generator.py:275]: easier to fix code than add another library dependency. General philosophy to date has been to use what is in Databricks runtime release or in preparing it - hence preference for specific versions of numpy, pandas, use of Sphinx etc.

That way, we can take advantage of the product testing performed during the DB runtime release to ensure compatibility of different versions of different libraries etc without having to test ourselves. I.e assumption is that if DB runtime team have validated version compatibility, assumption is that it works

File: dbldatagen/column_generation_spec.py (hunk 1 above)
Comment (ronanstokes-db) [reply dbldatagen/column_generation_spec.py:19]: re ISSUE TEMPLATE - fixed in separate checkin

File: dbldatagen/column_generation_spec.py (hunk 2 above)
Comment (ronanstokes-db) [reply dbldatagen/column_generation_spec.py:90]: Also - want to only use frameworks used by DB Runtime releases

**Comment by ronanstokes-db** [comment]
v0.3.1 is target release version for this feature

**Review by howardwu-db — CHANGES_REQUESTED** [review]
Resolve the warning issue about replacement of id columns logic

File: dbldatagen/data_generator.py [hunk 4]
```diff
@@ -693,6 +713,10 @@ def withColumn(self, colName, colType=StringType(), minValue=None, maxValue=None
         if baseColumn is not None:
             self._checkColumnOrColumnList(baseColumn, allowId=True)
 
+        if not noWarn and colName == DEFAULT_SEED_COLUMN and self._seedColumnName == DEFAULT_SEED_COLUMN:
+            self.logger.warning(f"Adding a column named '{colName}' overriding seed column '{self._seedColumnName}'")
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/data_generator.py]: message has been changed

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**